### PR TITLE
Added animefreak

### PR DESCRIPTION
**Pull request by Blatzar**
You might wanna add a setup.py and add it to pypi to make the installation easier.
Not sure if entering urls to cli works, just skimming the cli code I can't find it either, that makes testing the provider harder :(

For m3u8 I can reccomend looking at https://github.com/Blatzar/m3u8_downloader, minimal dependency wise and I have yet to find another good python m3u8 downloader without ffmpeg.

### CODE REVIEW

**Comment by justfoolingaround** [comment]
Oh, I'm sorry if it was difficult for you to use the added fetcher. The regex checking is done immediately in the search query prompt. That initial regex match will prevent the searcher to search from the site.

With HLS downloading, I'd like to inquire if your module checks the m3u8 url for more m3u8 url. There are streaming services that throw in bandwidth based list of m3u8 even in an m3u8. `ffmpeg`, as far as I know, does this and returns the does the job with the best m3u8 (that contains ts pulls).

I will soon add this project to PyPI when it becomes a little stable (that is, gets a better cli experience and maybe even integrations such as MyAnimeList).  

AnimeFreak does seem like a good site that does not contain any API calls and that at max, it requires one initializing call (basically getting the anime page and extracting episode links, just before the yielding starts) and one fetching call each (URL extraction from a episode link and immediate yield of it).

I'll keep this unmerged until AnimeFreak is fully functional, let me know if you need any sort of help with this.

**Comment by justfoolingaround** [comment]
Also, lol, the extraction process is more complicated as it seems. 

There is a hit & trial with random numbers (I wouldn't mind doing this if the even the streaming site does this) between (6~11) which is the (`"https://st%d.anime1.com/..." % random_number`) for the extracted streams.

image

I think you should use this regex for extraction for doing that hit and trial with ease:
```
(?:https?://)?st\d+\.anime1\.com/(?P<file>[^"]+)```

as every anime I've checked has the Anime1 server url.

I've tested your fork once and it kinda failed. I didn't want to interfere with any of your code lol as it's the last thing a developer wants.

**Comment by justfoolingaround** [comment]
Aight, all done. Let me know if there's anything. I'll merge this tomorrow.

**Comment by Blatzar** [comment]
> https://st%d.anime1.com/..." % random_number

True, but as it worked on the show I tested so I missed it :pensive:

File: core/providers/animefreak/__init__.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+def send_valid(session, stream_url):
+
+    with session.get(stream_url, stream=True) as response:
+        if response.ok:
+            return stream_url
+    
+    file = STREAM_URL_RE.search(stream_url).group('file')
+    
+    for trial in range(6, 12):
+        with session.get(url := BASE_STREAM_URL % (trial, file), stream=True) as response:
```
Comment (Blatzar) [inline core/providers/animefreak/__init__.py:20]: While the walrus operator is very nice to use, I think effort should be made to remove these as it's a 3.8+ feature. Many LTS distros are still on ~3.6.

**Comment by Blatzar** [comment]
> With HLS downloading, I'd like to inquire if your module checks the m3u8 url for more m3u8 url. There are streaming services that throw in bandwidth based list of m3u8 even in an m3u8. `ffmpeg`, as far as I know, does this and returns the does the job with the best m3u8 (that contains ts pulls).

Not supported, but fixing that shouldn't be very hard

**Comment by Blatzar** [comment]
> Oh, I'm sorry if it was difficult for you to use the added fetcher. The regex checking is done immediately in the search query prompt. 

You should probably add command parameters like 
```bash
python cli.py "https://twist.moe/a/one-piece"
```
to make that more intuitive, faster and scriptable

File: core/providers/animefreak/__init__.py (hunk 1 above)
Comment (justfoolingaround) [reply core/providers/animefreak/__init__.py:20]: Those operators are scattered all over the code. Oh boy, that's going to be a pain to remove. I'm sorry if they caused you trouble (you know, upgrading and stuff). I'll replace them as soon as I get enough time.

**Comment by justfoolingaround** [comment]
> > Oh, I'm sorry if it was difficult for you to use the added fetcher. The regex checking is done immediately in the search query prompt.
> 
> You should probably add command parameters like
> 
> ```shell
> python cli.py "https://twist.moe/a/one-piece"
> ```
> 
> to make that more intuitive, faster and scriptable

I think better alternative would be to add a message in the prompt itself. To make it more faster, yep, I think I'll resort to `argparse` so that users can configure the entire program just like that. Hence, not having any arguments to parse would allow the program to fallback to the current "cli".

**Pull request merged**